Template bug in `cholesky_decompose`

This bug exists in cmdstan 2.36 and 2.37.0-rc1

The following should compile but produces errors
```stan
data {
  int<lower=0> G;
  int<lower=0> C;
  matrix[G, C] A;
}
parameters {
  cholesky_factor_corr[G] L;
}
transformed parameters {
  matrix[C, C] L_c = 
    cholesky_decompose(A * multiply_lower_tri_self_transpose(L) * A');
}
model {
  L ~ lkj_corr_cholesky(4);
}
```

<details>

<summary><strong>Click to expand error message</strong></summary>

```
Compiling Stan program...
In file included from /var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f33110aeb24.hpp:1:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/src/stan/model/model_header.hpp:6:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math.hpp:19:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/rev.hpp:11:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/rev/constraint.hpp:24:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/constraint.hpp:4:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/fun.hpp:41:
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/fun/cholesky_decompose.hpp:39:7: error: no viable conversion from 'const LLT<Matrix<[3 * ...], 1, [2 * ...]>>' to 'LLT<Matrix<[3 * ...], (default) AutoAlign | ((_Rows == 1 && _Cols != 1) ? Eigen::RowMajor : (_Cols == 1 && _Rows != 1) ? Eigen::ColMajor : Eigen::ColMajor) aka 0, [2 * ...]>>'
   39 |       llt = m_eval.llt();
      |       ^     ~~~~~~~~~~~~
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f33110aeb24.hpp:140:21: note: in instantiation of function template specialization 'stan::math::cholesky_decompose<Eigen::Product<Eigen::Product<Eigen::Map<Eigen::Matrix<double, -1, -1>>, Eigen::Matrix<double, -1, -1>>, Eigen::Transpose<const Eigen::Map<Eigen::Matrix<double, -1, -1>>>>, nullptr, nullptr>' requested here
  140 |         stan::math::cholesky_decompose(
      |                     ^
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f33110aeb24.hpp:442:12: note: in instantiation of function template specialization 'test_error_model_namespace::test_error_model::log_prob_impl<false, false, Eigen::Matrix<double, -1, 1>, Eigen::Matrix<int, -1, 1>, nullptr, nullptr, 
nullptr>' requested here
  442 |     return log_prob_impl<propto__, jacobian__>(params_r, params_i, pstream);
      |            ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/src/stan/model/model_base_crtp.hpp:93:50: note: in instantiation of function template specialization 'test_error_model_namespace::test_error_model::log_prob<false, false, double>' requested here
   93 |     return static_cast<const M*>(this)->template log_prob<false, false, dou
ble>(
      |                                                  ^
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f33110aeb24.hpp:27:3: note: in instantiation of member function 'stan::model::model_base_crtp<test_error_model_namespace::test_error_model>::log_prob' requested here
   27 |   ~test_error_model() {}
      |   ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Core/util/ForwardDeclarations.h:280:55: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const LLT<PlainObject>' (aka 'const LLT<Eigen::Matrix<double, -1, -1, 1, -1, -1>>') to 'const LLT<Matrix<double, -1, -1>> &' for 1st argument
  280 | template<typename MatrixType, int UpLo = Lower> class LLT;
      |                                                       ^~~
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Core/util/ForwardDeclarations.h:280:55: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const LLT<PlainObject>' (aka 'const LLT<Eigen::Matrix<double, -1, -1, 1, -1, -1>>') to 'LLT<Matrix<double, -1, -1>> &&' for 1st argument
  280 | template<typename MatrixType, int UpLo = Lower> class LLT;
      |                                                       ^~~
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:101:14: note: explicit constructor is not a candidate
  101 |     explicit LLT(Index size) : m_matrix(size, size),
      |              ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:105:14: note: explicit constructor is not a candidate
  105 |     explicit LLT(const EigenBase<InputType>& matrix)
      |              ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:120:14: note: explicit constructor is not a candida
te
  120 |     explicit LLT(EigenBase<InputType>& matrix)
      |              ^

1 error generated.

make: *** [/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f33110aeb24.o] Error 1

Error: An error occured during compilation! See the message above for more information.
> mod <- cmdstan_model("test_error.stan")
Compiling Stan program...
In file included from /var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f331193cab23.hpp:1:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/src/stan/model/model_header.hpp:6:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math.hpp:19:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/rev.hpp:11:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/rev/constraint.hpp:24:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/constraint.hpp:4:
In file included from /Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/fun.hpp:41:
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/stan/math/prim/fun/cholesky_decompose.hpp:39:7: error: no viable conversion from 'const LLT<Matrix<[3 * ...], 1, [2 * ...]>>' to 'LLT<Matrix<[3 * ...], (default) AutoAlign | ((_Rows == 1 && _Cols != 1) ? Eigen::RowMajor : (_Cols == 1 && _Rows != 1) ? Eigen::ColMajor : Eigen::ColMajor) aka 0, [2 * ...]>>'
   39 |       llt = m_eval.llt();
      |       ^     ~~~~~~~~~~~~
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f331193cab23.hpp:153:21: note: in instantiation of function template specialization 'stan::math::cholesky_decompose<Eigen::Product<Eigen::Product<Eigen::Map<Eigen::Matrix<double, -1, -1>>, Eigen::Matrix<double, -1, -1>>, Eigen::Transpose<const Eigen::Map<Eigen::Matrix<double, -1, -1>>>>, nullptr, nullptr>' requested here
  153 |         stan::math::cholesky_decompose(
      |                     ^
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f331193cab23.hpp:479:12: note: in instantiation of function template specialization 'test_error_model_namespace::test_error_model::log_prob_impl<false, false, Eigen::Matrix<double, -1, 1>, Eigen::Matrix<int, -1, 1>, nullptr, nullpt
r, nullptr>' requested here
  479 |     return log_prob_impl<propto__, jacobian__>(params_r, params_i, pstream);
      |            ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/src/stan/model/model_base_crtp.hpp:93:50: note: in instantiation of function template specialization 'test_error_model_namespace::test_error_model::log_prob<false, false, double>' requested here
   93 |     return static_cast<const M*>(this)->template log_prob<false, false, double>(
|
      |                                                  ^
/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f331193cab23.hpp:30:3: note: in instantiation of member function 'stan::model::model_base_crtp<test_error_model_namespace::test_error_model>::log_prob' requested here
   30 |   ~test_error_model() {}
      |   ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Core/util/ForwardDeclarations.h:280:55: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const LLT<PlainObject>' (aka 'const LLT<Eigen::Matrix<double, -1, -1, 1, -1, -1>>') to 'const LLT<Matrix<double, -1, -1>> &' for 1st argument
  280 | template<typename MatrixType, int UpLo = Lower> class LLT;
      |                                                       ^~~
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Core/util/ForwardDeclarations.h:280:55: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const LLT<PlainObject>' (aka 'const LLT<Eigen::Matrix<double, -1, -1, 1, -1, -1>>') to 'LLT<Matrix<double, -1, -1>> &&' for 1st argument
  280 | template<typename MatrixType, int UpLo = Lower> class LLT;
      |                                                       ^~~
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:101:14: note: explicit constructor is not a candidate
  101 |     explicit LLT(Index size) : m_matrix(size, size),
      |              ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:105:14: note: explicit constructor is not a candidate
  105 |     explicit LLT(const EigenBase<InputType>& matrix)
      |              ^
/Users/sean.pinkney/.cmdstan/cmdstan-2.37.0-rc1/stan/lib/stan_math/lib/eigen_3.4.0/Eigen/src/Cholesky/LLT.h:120:14: note: explicit constructor is not a candidate
 
 120 |     explicit LLT(EigenBase<InputType>& matrix)
      |              ^

1 error generated.

make: *** [/var/folders/zh/nvq_p1td6c10x82ffzp9dns00000gq/T/RtmpFe7HAh/model-f331193cab23.o] Error 1

Error: An error occured during compilation! See the message above for more information.
```

</details>


By explicitly declaring a new matrix `S` that does the multiply does successfully compile 
```stan
data {
  int<lower=0> G;
  int<lower=0> C;
  matrix[G, C] A;
}
parameters {
  cholesky_factor_corr[G] L;
}
transformed parameters {
  matrix[C, C] S = A * multiply_lower_tri_self_transpose(L) * A';
  matrix[C, C] L_c =
    cholesky_decompose(S);
  // matrix[C, C] L_c = 
  //   cholesky_decompose(A * multiply_lower_tri_self_transpose(L) * A');
}
model {
  L ~ lkj_corr_cholesky(4);
}
```

--- a/stan/math/prim/fun/cholesky_decompose.hpp
+++ b/stan/math/prim/fun/cholesky_decompose.hpp
@@ -1,11 +1,9 @@
 #ifndef STAN_MATH_PRIM_FUN_CHOLESKY_DECOMPOSE_HPP
 #define STAN_MATH_PRIM_FUN_CHOLESKY_DECOMPOSE_HPP
 
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/fun/Eigen.hpp>
-
-#include <cmath>
 
 namespace stan {
 namespace math {
@@ -31,12 +29,11 @@ template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
 inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
                      EigMat::ColsAtCompileTime>
 cholesky_decompose(const EigMat& m) {
-  const eval_return_type_t<EigMat>& m_eval = m.eval();
+  using PlainMat = plain_type_t<EigMat>;
+  PlainMat m_eval = m;
   check_symmetric("cholesky_decompose", "m", m_eval);
   check_not_nan("cholesky_decompose", "m", m_eval);
-  Eigen::LLT<Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
-                           EigMat::ColsAtCompileTime>>
-      llt = m_eval.llt();
+  Eigen::LLT<PlainMat> llt = m_eval.llt();
   check_pos_definite("cholesky_decompose", "m", llt);
   return llt.matrixL();
 }

--- a/stan/math/prim/fun/cholesky_decompose.hpp
+++ b/stan/math/prim/fun/cholesky_decompose.hpp
@@ -1,9 +1,11 @@
 #ifndef STAN_MATH_PRIM_FUN_CHOLESKY_DECOMPOSE_HPP
 #define STAN_MATH_PRIM_FUN_CHOLESKY_DECOMPOSE_HPP
 
-#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+
+#include <cmath>
 
 namespace stan {
 namespace math {
@@ -26,14 +28,11 @@ namespace math {
  */
 template <typename EigMat, require_eigen_t<EigMat>* = nullptr,
           require_not_eigen_vt<is_var, EigMat>* = nullptr>
-inline Eigen::Matrix<value_type_t<EigMat>, EigMat::RowsAtCompileTime,
-                     EigMat::ColsAtCompileTime>
-cholesky_decompose(const EigMat& m) {
-  using PlainMat = plain_type_t<EigMat>;
-  PlainMat m_eval = m;
+inline plain_type_t<EigMat> cholesky_decompose(const EigMat& m) {
+  auto&& m_eval = to_ref(m);
   check_symmetric("cholesky_decompose", "m", m_eval);
   check_not_nan("cholesky_decompose", "m", m_eval);
-  Eigen::LLT<PlainMat> llt = m_eval.llt();
+  Eigen::LLT<plain_type_t<EigMat>> llt = m_eval.llt();
   check_pos_definite("cholesky_decompose", "m", llt);
   return llt.matrixL();
 }

--- a/test/unit/math/prim/fun/cholesky_decompose_test.cpp
+++ b/test/unit/math/prim/fun/cholesky_decompose_test.cpp
@@ -35,3 +35,18 @@ TEST(MathMatrixPrimMat, cholesky_decompose_exception) {
   EXPECT_THROW_MSG(stan::math::cholesky_decompose(m), std::domain_error,
                    "is not symmetric");
 }
+
+TEST(MathMatrixPrimMat, cholesky_decompose_expressions) {
+  // Test for https://github.com/stan-dev/math/issues/3198
+  stan::math::matrix_d A(2, 3);
+  A << 1, 2, 3, 4, 5, 6;
+
+  stan::math::vector_d L_u(3);
+  L_u << 1, 0, 0.5;
+
+  auto L = stan::math::cholesky_corr_constrain(L_u, 3);
+
+  EXPECT_NO_THROW(stan::math::cholesky_decompose(stan::math::multiply(
+      stan::math::multiply(A, stan::math::multiply_lower_tri_self_transpose(L)),
+      stan::math::transpose(A))));
+}

